crates/sui-graphql-rpc/src/zklogin_verify_signature.rs

```diff
@@ -0,0 +1,199 @@
+            IntentScope::try_from(intent_scope as u8)
+                .map_err(|_| Error::Client("Unsupported intent scope".to_string()))
+        }?;
+
+        // get current epoch from db.
+        let Some(curr_epoch) = Epoch::query(db, None, None).await? else {
+            return Err(Error::Internal(
+                "Cannot get current epoch from db".to_string(),
+            ));
+        };
+        let curr_epoch = curr_epoch.stored.epoch as u64;
+
+        // get chain id from db and determine zklogin_env.
+        let chain_id = ChainIdentifier::query(db).await?;
+        let zklogin_env = match chain_id == get_mainnet_chain_identifier()
+            || chain_id == get_testnet_chain_identifier()
+        {
+            true => ZkLoginEnv::Prod,
+            _ => ZkLoginEnv::Test,
+        };
+
+        // fetch on-chain JWKs from dynamic field of system object.
+        let df = DynamicField::query(
+            db,
+            SuiAddress::from_str("0x7")
+                .map_err(|_| Error::Internal("Invalid system object".to_string()))?,
+            None,
+            DynamicFieldName {
+                type_: ExactTypeFilter::from_str("u64")
+                    .map_err(|_| Error::Internal("Invalid df name type".to_string()))?,
+                bcs: Base64::from_str("AQAAAAAAAAA=")
+                    .map_err(|_| Error::Internal("Invalid df name bcs".to_string()))?,
+            },
+            DynamicFieldType::DynamicField,
+            None,
+        )
+        .await
+        .map_err(|e| Error::Internal(e.to_string()))?;
+
+        let binding = df.ok_or(Error::Internal("Cannot find df".to_string()))?;
+
+        let move_object = binding
+            .super_
+            .super_
+            .native_impl()
+            .ok_or(Error::Internal("Cannot pase object from df".to_string()))?
+            .data
+            .try_as_move()
+            .ok_or_else(|| Error::Internal("df not a Move object".to_string()))?;
+
+        let inner = bcs::from_bytes::<Field<u64, AuthenticatorStateInner>>(move_object.contents())
+            .map_err(|err| {
+                Error::Client(format!("Invalid bcs for AuthenticatorStateInner {}", err))
+            })?
+            .value;
+
+        // construct verify params with active jwks and zklogin_env.
+        let mut oidc_provider_jwks = ImHashMap::new();
+        for active_jwk in &inner.active_jwks {
+            let ActiveJwk { jwk_id, jwk, .. } = active_jwk;
+            match oidc_provider_jwks.entry(jwk_id.clone()) {
+                im::hashmap::Entry::Occupied(_) => {
+                    warn!("JWK with kid {:?} already exists", jwk_id);
+                }
+                im::hashmap::Entry::Vacant(entry) => {
+                    warn!("inserting JWK with kid: {:?}", jwk_id);
+                    entry.insert(jwk.clone());
+                }
+            }
+        }
+        let verify_params = VerifyParams::new(oidc_provider_jwks, vec![], zklogin_env, true, true);
```

Comment:
noob q: what is this doing?

Reply:
this is all the params that we need to read from onchain (ie. JWKs and env variables) to verify a zkLogin signature. see `zk.verify_authenticator` this is the same entry function that is used when executing the zklogin transaction.